### PR TITLE
TokenUIViewSet - NBOTQA-39

### PR DESCRIPTION
**Pull request by sirisha-bode**
# Closes 
https://networktocode.atlassian.net/browse/NBOTQA-39
# What's Changed

- Refactored Token model UI views to use NautobotUIViewSet and the UI Component Framework.
- Added TokenTable to display Token data in tabular format (Key, Description, Write Enabled, Created, Expires, Actions).
- Added TokenFilterForm to support search and filtering.
- Added TokenBulkEditForm to enable bulk update functionality.
- Added last_updated field to the Token model for audit tracking.
- Updated navigation to include Tokens under the Administration tab with appropriate permissions.
- Updated lookup logic to handle the historical "user" namespace mapping to "users".
- Updated tests to align with the new ViewSet-based implementation.

# Screenshots
| View            | Before                                                                                     | After                                                                                       |
|-----------------|-------------------------------------------------------------------------------------------|-------------------------------------------------------------------------------------------|
| List View       |image|image|
| Detail View     | |image|
| Create View     |imageimage|imageimage|
| Update View     |imageimage|imageimage|
| Delete View     |image|image|
| Bulk Edit Added | |image
| Bulk Edit Added view| |image
| Bulk Delete View | |imageimage|
| Change Log | |image

# TODO

- [ ] Explanation of Change(s)
- [ ] Added change log fragment(s) (for more information see [the documentation](https://docs.nautobot.com/projects/core/en/stable/development/core/#creating-changelog-fragments))
- [ ] Attached Screenshots, Payload Example
- [ ] Unit, Integration Tests
- [ ] Documentation Updates (when adding/changing features)
- [ ] Example App Updates (when adding/changing features)
- [ ] Outline Remaining Work, Constraints from Design

### CODE REVIEW

File: changes/8582.housekeeping [hunk 1]
```diff
@@ -0,0 +1 @@
+Refactored Token model related UI views to use `NautobotUIViewSet` and `UI Component Framework`.
```
Comment (Sridevi-NTC) [inline changes/8582.housekeeping]: Should this go under changes/8582.added since it’s a new implementation rather than a refactor?

File: nautobot/users/models.py [hunk 2]
```diff
@@ -248,6 +249,7 @@ class Token(BaseModel):
     documentation_static_path = "docs/user-guide/platform-functionality/users/token.html"
     natural_key_field_names = ["pk"]  # default would be `["key"]`, which is obviously not ideal!
     is_metadata_associable_model = False
+    is_saved_view_model = True
```
Comment (Sridevi-NTC) [inline nautobot/users/models.py:274]: Can we add changelog fragment for `is_saved_view_model` and `last_updated`?

File: nautobot/users/tests/test_views.py [hunk 3]
```diff
@@ -24,8 +24,8 @@ def test_change_password_enabled(self):
         """
         profile_response = self.client.get(reverse("user:profile"))
         preferences_response = self.client.get(reverse("user:preferences"))
-        api_tokens_response = self.client.get(reverse("user:token_list"))
```
Comment (Sridevi-NTC) [inline nautobot/users/tests/test_views.py:27]: Why were these lines of code removed?

**Comment by Sridevi-NTC** [comment]
Can you please add changelog screenshots with data?

**Comment by Sridevi-NTC** [comment]
Add datepicker for `created `and `expires `filters

Date picker in add token is not working as expected

File: nautobot/users/views.py [hunk 4]
```diff
@@ -325,6 +333,49 @@ def post(self, request):
...
+        return obj
+
+    def get_queryset(self):
+        """
+        Limit visibility to tokens owned by the authenticated user.
+        """
+        queryset = super().get_queryset()
+        if isinstance(self.request.user, AnonymousUser):
+            return queryset.none()
+        return queryset.filter(user=self.request.user)
```
Comment (Sridevi-NTC) [inline nautobot/users/views.py]: Why is this needed?

File: changes/8582.housekeeping (hunk 1 above)
Comment (sirisha-bode) [reply changes/8582.housekeeping]: Updated as requested, should be resolved, let me If I miss anything

File: nautobot/users/models.py (hunk 2 above)
Comment (sirisha-bode) [reply nautobot/users/models.py:274]: Updated as requested, should be resolved, let me If I miss anything

File: nautobot/users/views.py (hunk 4 above)
Comment (sirisha-bode) [reply nautobot/users/views.py]: Updated as requested, should be resolved, let me If I miss anything

File: nautobot/users/tests/test_views.py (hunk 3 above)
Comment (sirisha-bode) [reply nautobot/users/tests/test_views.py:27]: I removed those lines because this test is only for checking “Change Password” on Profile and Preferences pages.
The API Tokens page is different and does not always show that text, so it was causing unrelated test failures.

File: nautobot/project-static/js/forms.js [hunk 5]
```diff
@@ -54,6 +54,9 @@ function watchSourceFields(context, targetField, sourceFields, repopulate){
     sourceFields.forEach(function(sourceFieldName){
         const sourceFieldId = `id_${sourceFieldName}`;
         const sourceField = context.getElementById(sourceFieldId);
+        if (!sourceField) {
```
Comment (itdependsnetworks) [inline nautobot/project-static/js/forms.js]: Can you comment as to the why this is needed and ensure to describe when this is used, such as when the data is in such a way that this occurs.

Comment (sirisha-bode) [reply nautobot/project-static/js/forms.js]: Updated as requested, should be resolved, let me If I miss anything

Comment (itdependsnetworks) [reply nautobot/project-static/js/forms.js]: What is a sourcefield in this context? Is it the value of the form field on start? What is it about your use case that this is suddenly an issue?

Comment (sirisha-bode) [reply nautobot/project-static/js/forms.js]: In this code, sourceField is the actual form input element whose live value drives auto-population of the target field, and the null check is needed because shared initialization now runs on conditional form layouts where some source inputs may not be rendered.

File: nautobot/core/utils/lookup.py [hunk 6]
```diff
@@ -367,6 +367,11 @@ def get_model_for_view_name(view_name):
         app_label = app_label.replace("-api", "")
         delimiter = "-"
 
+    # Historical namespace for nautobot.users URLs is "user" (singular),
+    # but the Django app label is "users" (plural).
+    if app_label == "user":
```
Comment (glennmatthews) [inline nautobot/core/utils/lookup.py:372]: What's the code path where we hit this, out of curiosity?

File: nautobot/project-static/js/forms.js (hunk 5 above)
Comment (glennmatthews) [reply nautobot/project-static/js/forms.js]: Can you be more specific? Which form fields are necessitating this change? The added forms seem very straightforward to me nor do I see any changes to the form layouts/templates.

File: nautobot/users/forms.py [hunk 7]
```diff
@@ -46,6 +50,28 @@ class Meta:
         }
 
 
+class TokenFilterForm(NautobotFilterForm):
+    model = Token
+    q = forms.CharField(required=False, label="Search")
+    description = forms.CharField(required=False)
+    write_enabled = forms.NullBooleanField(required=False, widget=StaticSelect2(choices=BOOLEAN_WITH_BLANK_CHOICES))
+    created = NullableDateField(required=False, widget=DateTimePicker())
+    expires = NullableDateField(required=False, widget=DateTimePicker())
```
Comment (glennmatthews) [inline nautobot/users/forms.py]: These don't seem very useful as implemented since they're doing exact-match on a DateTimeField.

File: nautobot/users/models.py [hunk 8]
```diff
@@ -232,14 +232,15 @@ class Meta:
...
-class Token(BaseModel):
+class Token(BaseModel, ChangeLoggedModel):
     """
     An API token used for user authentication. This extends the stock model to allow each user to have multiple tokens.
     It also supports setting an expiration time and toggling write ability.
     """
 
     user = models.ForeignKey(to=settings.AUTH_USER_MODEL, on_delete=models.CASCADE, related_name="tokens")
     created = models.DateTimeField(auto_now_add=True)
+    last_updated = models.DateTimeField(auto_now=True)
```
Comment (glennmatthews) [inline nautobot/users/models.py]: These explicit declarations shouldn't be needed given that we're now using `ChangeLoggedModel`.

File: nautobot/users/models.py [hunk 9]
```diff
@@ -232,14 +232,15 @@ class Meta:
...
     expires = models.DateTimeField(blank=True, null=True)
     key = models.CharField(max_length=40, unique=True, validators=[MinLengthValidator(40)])
```
Comment (glennmatthews) [inline nautobot/users/models.py:267]: Please test and verify that the changelog entries created for Tokens do **NOT** include/expose the `key`!

File: nautobot/users/navigation.py [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+                            "users.view_token",
+                        ],
+                        buttons=(
+                            NavMenuAddButton(
+                                link="users:token_add",
+                                permissions=[
+                                    "users.add_token",
+                                ],
+                            ),
+                        ),
```
Comment (glennmatthews) [inline nautobot/users/navigation.py]: We don't use `buttons` any more in the Nav menu.

File: nautobot/users/tables.py [hunk 11]
```diff
@@ -0,0 +1,38 @@
+import django_tables2 as tables
+
+from nautobot.core.tables import BaseTable, BooleanColumn, ButtonsColumn, ToggleColumn
+
+from .models import Token
+
+
+class TokenTable(BaseTable):
+    pk = ToggleColumn()
+    key = tables.Column(linkify=True)
```
Comment (glennmatthews) [inline nautobot/users/tables.py]: Highly doubt that we should show `key` in the table view. In general it should be treated as a secret and not ever exposed in the UI to any user except its creator/owner, and maybe even then *only* on initial creation.

File: nautobot/users/tables.py [hunk 12]
```diff
@@ -0,0 +1,38 @@
+import django_tables2 as tables
+
+from nautobot.core.tables import BaseTable, BooleanColumn, ButtonsColumn, ToggleColumn
+
+from .models import Token
+
+
+class TokenTable(BaseTable):
+    pk = ToggleColumn()
```
Comment (glennmatthews) [inline nautobot/users/tables.py:13]: Probably should add a column showing the `user` for each token.

File: nautobot/users/tables.py [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+            "created",
+            "expires",
+            "actions",
+        )
+        default_columns = (
+            "pk",
+            "key",
+            "description",
+            "write_enabled",
+            "created",
```
Comment (glennmatthews) [inline nautobot/users/tables.py]: In general we don't usually include `created` or `last_updated` in `default_columns`.

File: nautobot/users/views.py [hunk 14]
```diff
@@ -318,6 +325,49 @@ def post(self, request):
...
+        """
+        Limit visibility to tokens owned by the authenticated user.
+        """
+        queryset = super().get_queryset()
+        if not self.request.user.is_authenticated:
+            return queryset.none()
+        return queryset.filter(user=self.request.user)
+
+
 class TokenListView(GenericView):
```
Comment (glennmatthews) [inline nautobot/users/views.py]: Going to remove these views as unused now?

File: nautobot/users/views.py [hunk 15]
```diff
@@ -318,6 +325,49 @@ def post(self, request):
...
+        return obj
+
+    def get_queryset(self):
+        """
+        Limit visibility to tokens owned by the authenticated user.
+        """
+        queryset = super().get_queryset()
+        if not self.request.user.is_authenticated:
+            return queryset.none()
+        return queryset.filter(user=self.request.user)
```
Comment (glennmatthews) [inline nautobot/users/views.py]: Let's make sure we think about both the end-user and the system administrator use cases for Tokens - not sure if that's in scope for this PR but admin users should be able to create/edit/delete Tokens for other users (but not view/edit the `key` values for existing tokens).

**Review by glennmatthews — COMMENTED** [review]
What's the plan for adding test coverage of this?

File: nautobot/project-static/js/forms.js (hunk 5 above)
Comment (sirisha-bode) [reply nautobot/project-static/js/forms.js]: The change wasn’t due to a form layout/template update. It was added because, in the Token add/edit form flow, the datetime picker field was not initializing reliably, so selecting a datetime could fail. The guard prevents shared form JS from breaking when a referenced source field isn’t present in that form context, which keeps the datetime field behavior stable.

**Comment by sirisha-bode** [comment]
> What's the plan for adding test coverage of this?

Added testcases

File: nautobot/core/utils/lookup.py (hunk 6 above)
Comment (sirisha-bode) [reply nautobot/core/utils/lookup.py:372]: We still have some old links that say user, but Django now uses users. This line simply translates old wording to the current one so everything keeps working without breaking older routes.

File: nautobot/users/forms.py (hunk 7 above)
Comment (sirisha-bode) [reply nautobot/users/forms.py]: Updated as requested, should be resolve, let me know if I missed anything

File: nautobot/users/models.py (hunk 8 above)
Comment (sirisha-bode) [reply nautobot/users/models.py]: Updated as requested, should be resolve, let me know if I missed anything

File: nautobot/users/models.py (hunk 9 above)
Comment (sirisha-bode) [reply nautobot/users/models.py:267]: Updated as requested, should be resolve, let me know if I missed anything

File: nautobot/users/navigation.py (hunk 10 above)
Comment (sirisha-bode) [reply nautobot/users/navigation.py]: Updated as requested, should be resolve, let me know if I missed anything

File: nautobot/users/tables.py (hunk 11 above)
Comment (sirisha-bode) [reply nautobot/users/tables.py]: Updated as requested, should be resolve, let me know if I missed anything

File: nautobot/users/tables.py (hunk 12 above)
Comment (sirisha-bode) [reply nautobot/users/tables.py:13]: Updated as requested, should be resolve, let me know if I missed anything

File: nautobot/users/tables.py (hunk 13 above)
Comment (sirisha-bode) [reply nautobot/users/tables.py]: Small question: if we hide key (for safety) and also stop showing created by default, which column should users click to open token details?
Description is optional, and User currently opens the user profile, not the token.

File: nautobot/users/views.py (hunk 14 above)
Comment (sirisha-bode) [reply nautobot/users/views.py]: Updated as requested, should be resolve, let me know if I missed anything

File: nautobot/users/views.py (hunk 15 above)
Comment (sirisha-bode) [reply nautobot/users/views.py]: Updated as requested, should be resolve, let me know if I missed anything

File: nautobot/users/migrations/0012_token_last_updated.py
Comment (glennmatthews) [inline nautobot/users/migrations/0012_token_last_updated.py:1]: Can these two migrations be squashed? It reads to me like 0012 shouldn't be needed at all.

File: nautobot/users/tests/test_views.py [hunk 16]
```diff
@@ -190,3 +189,165 @@ def test_timezone_change(self):
...
+        "description": "updated-via-ui-test",
+        "write_enabled": False,
+    }
+
+    bulk_edit_data = {
+        "description": "bulk-updated-token",
+    }
+
+    def _get_base_url(self):
+        return "user:token_{}"
```
Comment (glennmatthews) [inline nautobot/users/tests/test_views.py:213]: Not `users`?

File: nautobot/users/tests/test_views.py [hunk 17]
```diff
@@ -190,3 +189,165 @@ def test_timezone_change(self):
...
+
+        Token.objects.create(user=self.user, description="seed-token-1")
+        Token.objects.create(user=self.user, description="seed-token-2")
+        Token.objects.create(user=self.user, description="seed-token-3")
+
+    @override_settings(EXEMPT_VIEW_PERMISSIONS=["*"])
+    def test_get_object_anonymous(self):
+        self.client.logout()
+        response = self.client.get(self._get_queryset().first().get_absolute_url())
+        self.assertHttpStatus(response, 404)
```
Comment (glennmatthews) [inline nautobot/users/tests/test_views.py:234]: Why does this generic test need to be overridden?

File: nautobot/users/tests/test_views.py [hunk 18]
```diff
@@ -190,3 +189,165 @@ def test_timezone_change(self):
...
+
+    @override_settings(EXEMPT_VIEW_PERMISSIONS=["user:token_list"])
+    def test_list_objects_anonymous_with_exempt_permission_for_one_view_only(self):
```
Comment (glennmatthews) [inline nautobot/users/tests/test_views.py:237]: Can you add a docstring explaining this test? It seems strange to me that we're specifically marking this view as exempt but still expecting it to redirect to login.

File: nautobot/users/tests/test_views.py [hunk 19]
```diff
@@ -190,3 +189,165 @@ def test_timezone_change(self):
...
+        self.client.logout()
+        response = self.client.get(self._get_url("list"))
+        self.assertHttpStatus(response, 302)
+        self.assertIn("/login/", response.url)
+
+    @override_settings(EXEMPT_VIEW_PERMISSIONS=[])
+    def test_custom_actions(self):
+        self.skipTest("Token model does not support notes custom action.")
```
Comment (glennmatthews) [inline nautobot/users/tests/test_views.py]: Maybe it should?

File: nautobot/users/tests/test_views.py [hunk 20]
```diff
@@ -190,3 +189,165 @@ def test_timezone_change(self):
...
+
+    @override_settings(EXEMPT_VIEW_PERMISSIONS=[])
+    def test_filter_form_fields_are_working(self):
```
Comment (glennmatthews) [inline nautobot/users/tests/test_views.py]: Shouldn't this be covered by tests in `test_filters.py` instead?

File: nautobot/users/models.py [hunk 21]
```diff
@@ -248,6 +249,24 @@ class Token(BaseModel):
     documentation_static_path = "docs/user-guide/platform-functionality/users/token.html"
     natural_key_field_names = ["pk"]  # default would be `["key"]`, which is obviously not ideal!
     is_metadata_associable_model = False
+    is_saved_view_model = True
+
+    def to_objectchange(self, action, *, related_object=None, object_data_extra=None, object_data_exclude=None):
+        """Remove token key from changelog to prevent secret exposure."""
+        fields_to_exclude = ["key"]
+        if not object_data_exclude:
+            object_data_exclude = fields_to_exclude
```
Comment (glennmatthews) [inline nautobot/users/models.py]: This feels off to me, should it instead be something like:
```
if not object_data_exclude:
    object_data_exclude = []
object_data_exclude += fields_to_exclude
```

File: nautobot/users/tables.py (hunk 13 above)
Comment (glennmatthews) [reply nautobot/users/tables.py]: Might make sense to have this work like the Job list view, where we have `Detail` as a custom addition to the `actions` column?

File: nautobot/users/migrations/0012_token_last_updated.py
Comment (sirisha-bode) [reply nautobot/users/migrations/0012_token_last_updated.py:1]: Updated as requested, should be resolve, let me know if I missed anything

File: nautobot/users/tests/test_views.py (hunk 18 above)
Comment (sirisha-bode) [reply nautobot/users/tests/test_views.py:237]: Updated as requested, should be resolve, let me know if I missed anything

File: nautobot/users/tests/test_views.py (hunk 19 above)
Comment (sirisha-bode) [reply nautobot/users/tests/test_views.py]: Updated as requested, should be resolve, let me know if I missed anything

File: nautobot/users/tests/test_views.py (hunk 20 above)
Comment (sirisha-bode) [reply nautobot/users/tests/test_views.py]: Updated as requested, should be resolve, let me know if I missed anything

File: nautobot/users/models.py (hunk 21 above)
Comment (sirisha-bode) [reply nautobot/users/models.py]: Updated as requested, should be resolve, let me know if I missed anything

File: nautobot/users/tables.py (hunk 13 above)
Comment (sirisha-bode) [reply nautobot/users/tables.py]: Updated as requested, should be resolve, let me know if I missed anything

File: nautobot/users/tests/test_views.py (hunk 16 above)
Comment (sirisha-bode) [reply nautobot/users/tests/test_views.py:213]: The user (singular) namespace is intentional — it's defined as app_name = "user" in nautobot/users/urls.py. There's actually a # TODO comment in that file noting the inconsistency and flagging it for future deprecation in favor of users.

File: nautobot/users/tests/test_views.py (hunk 17 above)
Comment (sirisha-bode) [reply nautobot/users/tests/test_views.py:234]: This is overridden because the base generic test expects a 200 when EXEMPT_VIEW_PERMISSIONS=["*"] — i.e., anonymous users can access the object detail. For Token, we intentionally return 404 instead, even with all view permissions exempted, to prevent anonymous users from viewing any token detail. It's a security restriction specific to tokens, so the base behaviour needs to be overridden here.

File: nautobot/core/utils/lookup.py (hunk 6 above)
Comment (dsoftait) [reply nautobot/core/utils/lookup.py:372]: Can you give an example of code path where this situation is happen?

File: nautobot/project-static/js/forms.js [hunk 22]
```diff
@@ -78,7 +86,12 @@ function getSlugField(){
 
 function initializeAutoField(context, field, sourceFieldsAttrName, defaultMaxLength = 255, transformValue = null){
     // Get source fields and length values set as html attributes on given field
-    const sourceFields = field.getAttribute(sourceFieldsAttrName).split(" ");
+    const sourceFieldsAttr = field.getAttribute(sourceFieldsAttrName);
+    if (!sourceFieldsAttr) {
+        // Guard `id_key`/similar non-slug fields (e.g. API Token key) with no source attribute to avoid `.split(null)` crashes.
+        return;
+    }
+    const sourceFields = sourceFieldsAttr.split(" ");
```
Comment (dsoftait) [inline nautobot/project-static/js/forms.js]: not sure if I understand these changes in this file. Can you please explain which these changes are needed?

File: nautobot/users/tests/test_views.py [hunk 23]
```diff
@@ -111,8 +111,7 @@ def test_change_password_disabled(self):
             # Check UI
             profile_response = self.client.get(reverse("user:profile"))
             preferences_response = self.client.get(reverse("user:preferences"))
-            api_tokens_response = self.client.get(reverse("user:token_list"))
-            for response in [profile_response, preferences_response, api_tokens_response]:
+            for response in [profile_response, preferences_response]:
```
Comment (dsoftait) [inline nautobot/users/tests/test_views.py:115]: What exactly changed that we had to remove `api_tokens_response`? I saw the comment above that there was a bug, but shouldn't we fix it instead of removing `api_tokens_response`?

File: nautobot/users/models.py (hunk 9 above)
Comment (dsoftait) [reply nautobot/users/models.py:267]: Can you point me to a test that checks this?

File: nautobot/core/utils/lookup.py (hunk 6 above)
Comment (sirisha-bode) [reply nautobot/core/utils/lookup.py:372]: I identified that the issue was caused by the Saved View storing view = "user:token_list" while the actual Django app label is users. Because of this, apps.get_model("user", "token") returned None, which later caused the 500 error.

I fixed it by adding a mapping from "user" to "users" so the Token model resolves correctly and Update Current now works as expected.

File: nautobot/users/models.py (hunk 9 above)
Comment (sirisha-bode) [reply nautobot/users/models.py:267]: I removed that implementation based on the earlier suggestion. Currently, I only added ChangeLoggedModel support for the last_updated field, so there is no separate test verifying that the Token key is excluded from changelog entries.

File: nautobot/users/tests/test_views.py (hunk 23 above)
Comment (sirisha-bode) [reply nautobot/users/tests/test_views.py:115]: Previously, the API Tokens page used users/base.html, which included the Profile/Preferences/Change Password navigation, so the test was checking for the "Change Password" link.

After migrating to TokenUIViewSet, the page became a standard ObjectListView and no longer includes that user-profile navigation. Because of this, the "Change Password" link is no longer expected on the page.

That’s why I removed only the API Tokens response from those assertions, while keeping the Profile/Preferences checks unchanged.

File: nautobot/project-static/js/forms.js (hunk 22 above)
Comment (sirisha-bode) [reply nautobot/project-static/js/forms.js]: Sure. getSlugField() in forms.js first looks for #id_slug, and if that's missing it falls back to #id_key. That fallback was added for CustomField and Relationship, whose key is a Nautobot SlugField and always carries a slug-source attribute.

Token also has a key field, but it's the API secret — a plain CharField with no slug-source. So on the Token add/edit page, field.getAttribute('slug-source').split(" ") becomes null.split(" ") and throws, which halts jsify_form and breaks flatpickr (Expires calendar), select2, checkboxes, etc.

The guard early-returns when there's no slug-source to read — nothing to wire up in that case. To reproduce on develop: open /user/api-tokens/add/ and the console shows Cannot read properties of null (reading 'split'), and the Expires calendar doesn't open.

File: nautobot/users/tests/test_views.py (hunk 16 above)
Comment (sirisha-bode) [reply nautobot/users/tests/test_views.py:213]: check django to see alias as possible?

File: nautobot/core/utils/lookup.py (hunk 6 above)
Comment (dsoftait) [reply nautobot/core/utils/lookup.py:372]: ok, can we cover this corner case in unittest?

File: nautobot/users/views.py (hunk 15 above)
Comment (dsoftait) [reply nautobot/users/views.py]: I couldn't find tests regarding this scenario. I mean I see that we check if `key` it's not present, but maybe worth it to have tests which check that
- admin users should be able to create/edit/delete Tokens for other users (but not view/edit the key values for existing tokens)
- not admin users shouldn't be able create/edit/delete Tokens

Comment (bennybates) [reply nautobot/users/views.py]: @sirisha-bode can you add additional tests her per Daria's comment?

File: nautobot/users/tables.py [hunk 24]
```diff
@@ -5,7 +5,7 @@
 from .models import Token
 
 _actions_template = """
-<li><a href="{% url 'user:token' pk=record.pk %}" class="dropdown-item"><span class="mdi mdi-information-outline" aria-hidden="true"></span>Details</a></li>
+<li><a href="{% url 'users:token' pk=record.pk %}" class="dropdown-item"><span class="mdi mdi-information-outline" aria-hidden="true"></span>Details</a></li>
```
Comment (dsoftait) [inline nautobot/users/tables.py]: If [here](https://github.com/nautobot/nautobot/pull/8582/changes/27d167055da5e4dfbebfff3c8a24adfe4259f99e#diff-4eaefccfac44e58e8fed929ab5c467214a750029cd67481c5636b0c6bbeb552fL212) you define `user:token` so why here do you change to `users:token`?

File: nautobot/project-static/js/forms.js (hunk 22 above)
Comment (glennmatthews) [reply nautobot/project-static/js/forms.js]: Should we extract this fix to a separate/additional PR to `develop` since it's an existing bug?

**Review by dsoftait — APPROVED** [review]
Right now looks good to me.So if the manual tests confirm this, then it's good to go

File: nautobot/users/tests/test_views.py [hunk 25]
```diff
@@ -190,3 +189,225 @@ def test_timezone_change(self):
...
+        self.add_permissions("users.view_token")
+        own_token = Token.objects.create(user=self.user, description="own-token-visible")
+        Token.objects.create(user=self.other_user, description="other-token-hidden")
+
+        response = self.client.get(reverse("user:token_list"), headers={"HX-Request": "true"})
+        self.assertHttpStatus(response, 200)
+        self.assertBodyContains(response, own_token.description)
+        self.assertNotIn("other-token-hidden", response.content.decode(response.charset))
+
+    def test_staff_list_shows_all_tokens(self):
```
Comment (glennmatthews) [inline nautobot/users/tests/test_views.py:292]: Probably should add some table tests asserting that the list view/table does *not* show the `key` as an enable-able column.

File: nautobot/users/tests/test_views.py (hunk 17 above)
Comment (glennmatthews) [reply nautobot/users/tests/test_views.py:234]: Can you add a docstring with the above information?

File: nautobot/users/tests/test_views.py [hunk 26]
```diff
@@ -190,3 +189,225 @@ def test_timezone_change(self):
...
+        the token list endpoint is exempted from auth (for cases like public listing with
+        filtering), anonymous clients still should not be allowed to view a specific token
+        object via detail URL, so redirect-to-login is expected here.
+        """
+        self.client.logout()
+        response = self.client.get(self._get_url("list"))
+        self.assertHttpStatus(response, 302)
+        self.assertIn("/login/", response.url)
+
+    def test_token_add_form_shows_key_field(self):
```
Comment (glennmatthews) [inline nautobot/users/tests/test_views.py:250]: I don't see any tests about the visibility of the `key` field in the token detail view, or which users can/cannot access the detail view for a given token. Probably need to add those?

File: nautobot/users/views.py [hunk 27]
```diff
@@ -318,121 +324,48 @@ def post(self, request):
...
+        return super().form_save(form, **kwargs)
+
+    def get_queryset(self):
+        """Allow staff/superusers to manage all tokens; others are limited to their own."""
+        queryset = super().get_queryset()
+        if not self.request.user.is_authenticated:
+            return queryset.none()
+        if self.request.user.is_staff or self.request.user.is_superuser:
+            return queryset
+        return queryset.filter(user=self.request.user)
```
Comment (glennmatthews) [inline nautobot/users/views.py]: Thinking about this more, I think we should look to move this logic *into* `Token.objects.restrict()` so that the same access restrictions apply whether the Token is accessed via a view, REST/GraphQL, or via an appropriate ORM call (such as in changelog access checks, #8694)

File: nautobot/users/models.py [hunk 28]
```diff
@@ -232,12 +235,32 @@ class Meta:
...
+    users only see their own. Staff/superusers can manage all tokens (e.g. to revoke a leaked
+    user's key). Centralizing this here means the same rule applies to UI views, REST/GraphQL,
+    and ORM call sites such as changelog access checks.
+    """
+
+    def restrict(self, user, action="view"):
+        if not user.is_authenticated:
+            return self.none()
+        qs = super().restrict(user, action)
+        if not (user.is_staff or user.is_superuser):
```
Comment (glennmatthews) [inline nautobot/users/models.py]: Let's change all of these cases to just `is_staff` - a superuser who is not staff should not have access to other users' tokens.

File: changes/8582.housekeeping [hunk 29]
```diff
@@ -0,0 +1 @@
+Added null guards to form initialization JavaScript so partial form layouts do not crash, mapped the "user" URL namespace to the "users" app label in `get_model_for_view_name`, and renamed `assertQuerysetEqualAndNotEmpty` call sites to `assertQuerySetEqualAndNotEmpty`.
```
Comment (glennmatthews) [inline changes/8582.housekeeping]: Since the JS fix will be introduced upstream (#8972) before this code ever makes its way to a release, we can skip mentioning it here. Also I don't see any changes in this PR about assertQuerySetEqualAndNotEmpty?

```suggestion
Mapped the "user" URL namespace to the "users" app label in `get_model_for_view_name`.
```

File: nautobot/users/api/serializers.py [hunk 30]
```diff
@@ -68,19 +69,40 @@ class Meta:
...
+        raw_user_input = data.get("user") if isinstance(data, dict) else None
+
+        validated = super().to_internal_value(data)
+
+        if "key" not in validated and not self.instance:
+            validated["key"] = Token.generate_key()
+
+        request_user = self.context["request"].user
+
+        # Mirror TokenUIViewSet.form_save: staff/superusers may explicitly assign ownership;
```
Comment (glennmatthews) [inline nautobot/users/api/serializers.py]: If we're going to do this (I'm not 100% convinced it's desirable) a cleaner/more correct fix might be to remove the `exclude=["user"]` above, mark `user` as optional, and raise an error if a non-staff user tries to set `user` to anything other than their own self --  instead of messing around with `raw_user_input`?

Comment (glennmatthews) [reply nautobot/users/api/serializers.py]: I don't think I see a valid use case for reassigning an existing Token to a different user account either via REST API or via UI - user assignment should IMHO *only* be done at Token creation time.

Comment (glennmatthews) [reply nautobot/users/api/serializers.py]: Actually - we probably shouldn't allow changing the `key` on an existing Token either.

Comment (glennmatthews) [reply nautobot/users/api/serializers.py]: For that matter, now that staff users can view other users' tokens, the REST API should probably treat `key` as write-only and *not* expose existing key values in response to a GET.

File: nautobot/users/forms.py [hunk 31]
```diff
@@ -32,10 +37,19 @@ class TokenForm(BootstrapMixin, forms.ModelForm):
         required=False,
         help_text="If no key is provided, one will be generated automatically.",
     )
+    user = forms.ModelChoiceField(queryset=get_user_model().objects.all(), required=False)
```
Comment (glennmatthews) [inline nautobot/users/forms.py]: 1) Can/should this be a `DynamicModelChoiceField` instead?
2) This should be hidden for non-staff users since they cannot meaningfully pick the user value
3) As with `key`, we should (IMO) disallow changing the `user` of an existing Token.

File: nautobot/users/tables.py [hunk 32]
```diff
@@ -0,0 +1,41 @@
+import django_tables2 as tables
+
+from nautobot.core.tables import BaseTable, BooleanColumn, ButtonsColumn, ToggleColumn
+
+from .models import Token
+
+_actions_template = """
+<li><a href="{% url 'user:token' pk=record.pk %}" class="dropdown-item"><span class="mdi mdi-information-outline" aria-hidden="true"></span>Details</a></li>
+"""
```
Comment (glennmatthews) [inline nautobot/users/tables.py:9]: Just a note/reminder that #8610 will remove the need for this.

File: nautobot/users/views.py [hunk 33]
```diff
@@ -19,19 +19,25 @@
...
 from nautobot.core.events import publish_event
-from nautobot.core.forms import ConfirmationForm
+from nautobot.core.ui import object_detail
+from nautobot.core.ui.choices import SectionChoices
 from nautobot.core.ui.titles import Titles
 from nautobot.core.views.generic import GenericView
+from nautobot.core.views.viewsets import NautobotUIViewSet
 from nautobot.users.utils import serialize_user_without_config_and_views
 
 from ..core.views.mixins import GetReturnURLMixin
```
Comment (glennmatthews) [inline nautobot/users/views.py]: Mind changing this to `from nautobot.core.views.mixins` while you're here?

**Comment by sirisha-bode** [comment]
# What's Changed (additions)

- Restricted access to staff users only — dropped `is_superuser` from the gate per reviewer feedback. A superuser-but-not-staff user is treated like a regular user.
- Made `user` and `key` immutable on existing tokens (both REST and UI). Only `description`, `write_enabled`, and `expires` can be edited after creation.
- `TokenSerializer`: `key` is now `write_only` and revealed only in the create response. List, retrieve, update, and CSV-export responses never include `key`.
- TokenForm: `user` switched to `DynamicModelChoiceField` (search-backed picker) and hidden from non-staff users on the add form.
- Added/updated tests for the new restrict rule, field immutability, key write-only behavior, and role-based form visibility.

# Role-Based Form Behavior

The `User` and `Key` fields are only meaningful at creation time. Staff users can assign tokens to other users on the add form; non-staff users always create tokens for themselves and never see the picker. On the edit form, both fields are hidden for **all** roles so existing tokens cannot be silently reassigned or have their key changed.

## Add Form by User Role

| Role | `User` field | Screenshot |
|------|--------------|------------|
| Regular user (no staff, no superuser) | Hidden — owner is forced to self | image |
| Superuser **but not** staff | Hidden — proves the new `is_staff`-only rule | image |
| Staff user | Shown as a `DynamicModelChoiceField` | image |

## Edit Form (all roles)

`User` and `Key` are removed from the form. Only `Description`, `Write Enabled`, and `Expires` are editable.

image

**Comment by itdependsnetworks** [comment]
Closing until we pick this back up at a future date
